symengine/polys/upolybase.h

```diff
@@ -280,13 +282,120 @@ class UIntPolyBase : public UPolyBase<Container, Poly>
     {
     }
 
+    // returns degree of the poly
     inline unsigned int get_degree() const
     {
         return this->poly_.degree();
     }
-
+    // return coefficient of degree 'i'
     virtual integer_class get_coeff(unsigned int i) const = 0;
+    // return value of poly when ealudated at `x`
     virtual integer_class eval(const integer_class &x) const = 0;
+    // return `degree` + 1. `0` returned for zero poly.
+    virtual unsigned int size() const = 0;
+
+    RCP<const Basic> as_symbolic() const
+    {
+        auto it = (static_cast<const Poly &>(*this)).begin();
+        auto end = (static_cast<const Poly &>(*this)).end();
+
+        vec_basic args;
+        for (; it != end; ++it) {
+            if (it->first == 0) {
+                args.push_back(integer(it->second));
+            } else if (it->first == 1) {
+                if (it->second == 1) {
+                    args.push_back(this->var_);
+                } else {
+                    args.push_back(Mul::from_dict(integer(it->second),
+                                                  {{this->var_, one}}));
+                }
+            } else {
+                if (it->second == 1) {
+                    args.push_back(pow(this->var_, integer(it->first)));
+                } else {
+                    args.push_back(
+                        Mul::from_dict(integer(it->second),
+                                       {{this->var_, integer(it->first)}}));
+                }
+            }
+        }
+        return SymEngine::add(args);
+    }
+};
+
+template <typename T>
+class ContainerBaseIter
+{
+protected:
+    RCP<const T> ptr_;
+    long i_;
+
+public:
+    ContainerBaseIter(RCP<const T> ptr, long x) : ptr_{ptr}, i_{x}
+    {
+    }
+
+    bool operator==(const ContainerBaseIter &rhs)
+    {
+        return (ptr_ == rhs.ptr_) and (i_ == rhs.i_);
+    }
+
+    bool operator!=(const ContainerBaseIter &rhs)
+    {
+        return not(*this == rhs);
+    }
+
+    std::pair<long, integer_class> operator*()
+    {
+        return std::make_pair(i_, ptr_->get_coeff(i_));
+    }
+
+    std::shared_ptr<std::pair<long, integer_class>> operator->()
+    {
+        return std::make_shared<std::pair<long, integer_class>>(
+            i_, ptr_->get_coeff(i_));
+    }
+};
+
+template <typename T>
+class ContainerForIter : public ContainerBaseIter<T>
+{
+public:
+    ContainerForIter(RCP<const T> ptr, long x) : ContainerBaseIter<T>(ptr, x)
+    {
+    }
+
+    ContainerForIter operator++()
+    {
+        this->i_++;
```

Comment:
We do I think. Compiler does not know `i_` lies in this class. (Template class derived from template class)